### PR TITLE
Fix events

### PR DESCRIPTION
**Pull request by jasonyihk**
use server.ext instead to be compatible with hapijs 16.x, 17.x & 18.x

### CODE REVIEW

File: packages/hapi/modules/plugin/plugin.js [hunk 1]
```diff
@@ -37,13 +37,13 @@ const createPlugin = ({ options } = {}) => {
   const plugin = {
     name: pkg.name,
     version: pkg.version,
-    register(server) {
+    register(server, o, done) {
```
Comment (tdeekens) [inline packages/hapi/modules/plugin/plugin.js]: Please check master. This is defaulted now as I can see no reference on hapi passing it in. Can you clarify?

File: packages/hapi/modules/plugin/plugin.js [hunk 2]
```diff
@@ -37,13 +37,13 @@ const createPlugin = ({ options } = {}) => {
...
       server.ext('onRequest', (request, h) => {
         request.promster = { start: process.hrtime() };
         return h.continue;
       });
 
-      server.events.on('response', request => {
```
Comment (tdeekens) [inline packages/hapi/modules/plugin/plugin.js]: I added a version check in later versions. Please check master if this does the trick for you.

File: packages/hapi/modules/plugin/plugin.js [hunk 3]
```diff
@@ -38,16 +37,13 @@ const createPlugin = ({ options } = {}) => {
...
-      const onRequestHandler = (request, h) => {
+    register(server, o, done) {
+      server.ext('onRequest', (request, h) => {
         request.promster = { start: process.hrtime() };
         return h.continue;
-      };
+      });
 
-      const onResponseHandler = request => {
+      server.ext('onPreResponse', (request, h) => {
```
Comment (tdeekens) [inline packages/hapi/modules/plugin/plugin.js]: Please check again with `master`. I now compare versions of Hapi to pick this API if < 17.0.x.

File: packages/hapi/modules/plugin/plugin.js (hunk 1 above)
Comment (tdeekens) [reply packages/hapi/modules/plugin/plugin.js]: This will break the plugin. `done` is not passed and invoking it below makes any server fail to start.

**Comment by jasonyihk** [comment]
would be simple just switch to server.ext('onPreResponse', without ay version dependency?

**Comment by tdeekens** [comment]
Thanks for getting back. Can you elaborate how this check

```js
const areServerEventsSupported = Boolean(
   server.version && semiver(server.version, '17.0.x')
);

// ... more

// NOTE: This version detection allows us to graceully support
// both new and old Hapi APIs.
if (areServerEventsSupported) {
    server.events.on('request', onRequestHandler);
    server.events.on('response', onResponseHandler);
} else {
    server.ext('onRequest', onRequestHandler);
    server.ext('onPreResponse', onResponseHandler);
}
```

does not allow us to be backwards compatible by using `server.ext` with versions smaller 17?

File: packages/hapi/modules/plugin/plugin.js (hunk 2 above)
Comment (jasonyihk) [reply packages/hapi/modules/plugin/plugin.js]: good catch. updated PR with checking on function done().

**Comment by jasonyihk** [comment]
debug log:

> server.version: 16.4.3
> semiver(server.version, '17.0.x'): -1
> areServerEventsSupported: true
> Boolean('16.4.3' && -1) : true

**Comment by tdeekens** [comment]
Thanks. I guess that we can fix.

**Comment by tdeekens** [comment]
How about https://github.com/tdeekens/promster/pull/149. Would that work for you. I also added tests for the cases.

**Comment by tdeekens** [comment]
Can you please try if `@promster/hapi@2.2.4` works for you. Thanks.

**Pull request closed without merging**